beacon-chain/blockchain/metrics.go

```diff
@@ -210,14 +221,31 @@ func reportEpochMetrics(ctx context.Context, postState, headState state.BeaconSt
 	beaconFinalizedRoot.Set(float64(bytesutil.ToLowInt64(postState.FinalizedCheckpoint().Root)))
 	currentEth1DataDepositCount.Set(float64(postState.Eth1Data().DepositCount))
 
-	// Validator participation should be viewed on the canonical chain.
-	v, b, err := precompute.New(ctx, headState)
-	if err != nil {
-		return err
-	}
-	_, b, err = precompute.ProcessAttestations(ctx, headState, v, b)
-	if err != nil {
-		return err
+	var b *precompute.Balance
+	var v []*precompute.Validator
+	var err error
+	switch headState.Version() {
+	case version.Phase0:
+		// Validator participation should be viewed on the canonical chain.
+		v, b, err = precompute.New(ctx, headState)
+		if err != nil {
+			return err
+		}
+		_, b, err = precompute.ProcessAttestations(ctx, headState, v, b)
+		if err != nil {
+			return err
+		}
+	case version.Altair:
+		v, b, err = altair.InitializeEpochValidators(ctx, headState)
+		if err != nil {
+			return err
+		}
+		_, b, err = altair.ProcessEpochParticipation(ctx, headState, b, v)
+		if err != nil {
+			return err
+		}
+	default:
+		return errors.Errorf("invalid state type provided: %T", headState.InnerStateUnsafe())
```

Comment:
> Why can a BeaconState get to this method with an invalid Version? Could there be some validation in the block processing that ensures this as an invariant?

This should be an impossible error, we have the default case as a defensive check if in case the impossible happens for some reason. Otherwise the head state would be nil and it would panic further down. 

> Also, what do you think about this idea -- instead of only providing access to the integer version via BeaconState.Version(), which requires code that wants balances to deal with fork-specific knowledge of the processing methods to call w/ headState and so forth, Version() could give you an interface type that could encapsulate that domain knowledge and offer a method like GetBalances that could be used by this metrics code?

This would be an interesting approach, we would have to tie it up with the core methods maybe ? And provide some sort of interface that abstracts the core functions per state so that the switch stuff here can be removed. Although I do not have a good idea mentally on how to make this work.